fairseq2n/src/fairseq2n/data/multi_data_source.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/data/multi_data_source.h"
+
+namespace fairseq2n::detail {
+
+multi_data_source::multi_data_source(std::vector<data_pipeline> &&pipelines, index_generator_fn &&index_gen_fn, bool stop_at_shortest)
+    : pipelines_(std::move(pipelines)), next_index_gen_(std::move(index_gen_fn)), stop_at_shortest_(stop_at_shortest)
+{
+    is_epoch_done_ = std::vector<bool>(pipelines_.size(), false);
+    buffer_ = std::vector<std::optional<data>>(pipelines_.size(), std::nullopt);
+}
+
+std::optional<data>
+multi_data_source::next()
+{
+    if (stop_at_shortest_) // with this flag on, the operator is a simple iterator
+        return pipelines_[next_index_gen_()].next();
+
+    // One or more data pipelines might be empty, so we have to keep looping
+    std::optional<data> output{};
+    while (!output && !eod()) {
```

Comment:
Neat! This is definitely a much cleaner implementation than mine in `round_robin`.